### PR TITLE
[SPARK-35339][PYTHON] Improve unit tests for data-type-based basic operations

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?

Improve unit tests for data-type-based basic operations by:
- removing redundant test cases
- adding `astype` test for ExtensionDtypes

### Why are the changes needed?

Some test cases for basic operations are duplicated after introducing data-type-based basic operations. The PR is proposed to remove redundant test cases.
`astype` is not tested for ExtensionDtypes, which will be adjusted in this PR as well.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests.

Keyword: SPARK-35337

### CODE REVIEW

**Comment by xinrong-databricks** [comment]
CC @ueshin @HyukjinKwon @itholic

**Comment by xinrong-databricks** [comment]
ERROR [4.809s]: test_await_termination_or_timeout (pyspark.streaming.tests.test_context.StreamingContextTests)
--------------------------------------------------------------------

Retriggering the test

**Comment by ueshin** [comment]
Jenkins, retest this please.

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py [hunk 1]
```diff
@@ -572,10 +578,28 @@ def test_isnull(self):
...
+                    Float32Dtype(),
+                    Float64Dtype(),
+                ]
+
+                if is_extension_float_dtype:
+                    self.assert_eq([1.0, 0.0, np.nan], self.psser.astype(dtype).tolist())
+                else:
+                    self.check_extension(pser.astype(dtype), psser.astype(dtype))
+            else:
+                self.check_extension(pser.astype(dtype), psser.astype(dtype))
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py:602]: How about:

```suggestion
            if dtype in self.fractional_extension_dtypes:
                self.assert_eq([1.0, 0.0, np.nan], self.psser.astype(dtype).tolist())
            else:
                self.check_extension(pser.astype(dtype), psser.astype(dtype))
```

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py [hunk 2]
```diff
@@ -572,10 +578,28 @@ def test_isnull(self):
...
+            if extension_float_dtypes_available:
+                is_extension_float_dtype = dtype in [
+                    "Float32",
+                    "Float64",
+                    Float32Dtype(),
+                    Float64Dtype(),
+                ]
+
+                if is_extension_float_dtype:
+                    self.assert_eq([1.0, 0.0, np.nan], self.psser.astype(dtype).tolist())
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py:598]: Could you add a comment why we need this?

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py [hunk 3]
```diff
@@ -572,10 +578,28 @@ def test_isnull(self):
     def test_astype(self):
         pser = self.pser
         psser = self.psser
+
+        # [True, False, <NA>] is returned in pandas
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py:582]: I guess we should follow the pandas' behavior?
Could you file a JIRA ticket to track this and put the JIRA ID like `# TODO(SPARK-xxxxx): [True, ...`, or address here?

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py (hunk 1 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py]: Better approach! Modified.

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py (hunk 2 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py]: Sure added.

File: python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py (hunk 3 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/data_type_ops/test_boolean_ops.py]: Makes sense, SPARK-35976 is created and commented here.

File: python/pyspark/pandas/tests/data_type_ops/test_string_ops.py [hunk 4]
```diff
@@ -217,6 +220,20 @@ def test_from_to_pandas(self):
     def test_isnull(self):
         self.assert_eq(self.pser.isnull(), self.psser.isnull())
 
+    def test_astype(self):
+        pser = self.pser
+        psser = self.psser
+
+        # [x, y, z, <NA>] is returned in pandas
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_string_ops.py:227]: Shall we add `TODO(SPARK-35976):` here as well?

Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/data_type_ops/test_string_ops.py]: For sure, added.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Comment by xinrong-databricks** [comment]
Thank you @ueshin !

**Pull request closed without merging**